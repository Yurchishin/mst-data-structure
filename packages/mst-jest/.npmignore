__test__
src
.babelrc
.eslintignore
.eslintrc
coverage
.gitignore
.npmignore
jest.config.js
rollup.config.js