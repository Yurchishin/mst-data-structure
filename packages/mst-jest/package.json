{
  "name": "@mst-ds/mst-jest",
  "version": "0.0.7",
  "description": "> TODO: description",
  "author": "Yurchishin",
  "homepage": "https://github.com/Yurchishin/mst-data-structure",
  "license": "MIT",
  "main": "lib/mst-jest.js",
  "umd:main": "lib/mst-jest.umd.js",
  "module": "lib/mst-jest.module.js",
  "browser": {
    "./lib/mst-jest.js": "./lib/mst-jest.js",
    "./lib/mst-jest.module.js": "./lib/mst-jest.module.js"
  },
  "unpkg": "lib/mst-jest.umd.min.js",
  "jsnext:main": "lib/mst-jest.module.js",
  "react-native": "lib/mst-jest.module.js",
  "directories": {
    "lib": "lib"
  },
  "files": [
    "lib/"
  ],
  "scripts": {
    "crean": "rm -rf lib/",
    "build": "rollup -c",
    "lint": "eslint --ext .ts src",
    "lint:fix": "eslint --fix --ext .js src"
  },
  "devDependencies": {
    "eslint": "^7.10.0",
    "mobx": "^5.15.4",
    "mobx-state-tree": "3.17.2",
    "rollup": "^2.21.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-filesize": "^9.0.2",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-replace": "^2.2.0",
    "rollup-plugin-terser": "^6.1.0"
  },
  "peerDependencies": {
    "mobx": ">=4.8.0 <5.0.0 || >=5.8.0 <6.0.0",
    "mobx-state-tree": "3.17.2"
  },
  "keywords": [
    "mobx",
    "mobx-state-tree",
    "reactive",
    "state management",
    "data structure"
  ],
  "gitHead": "38957bca5fdad6f33b29dc295ae6d60a4063cc69",
  "publishConfig": {
    "access": "public"
  }
}
