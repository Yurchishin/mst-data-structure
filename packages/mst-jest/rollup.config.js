import { baseConfig } from "../../rollup.base-config"

const config = (outFile, format, mode) =>
    baseConfig({
        outFile,
        format,
        mode,

        input: "./src/index.js",
        globals: {
            mobx: "mobx",
            "mobx-state-tree": "mobxStateTree",
        },
        umdName: "mstJest",
        external: ["mobx", "mobx-state-tree"],
    })

export default [
    config("mst-jest.js", "cjs", "development"),
    config("mst-jest.min.js", "cjs", "production"),

    config("mst-jest.umd.js", "umd", "development"),
    config("mst-jest.umd.min.js", "umd", "production"),

    config("mst-jest.module.js", "es", "development"),
]