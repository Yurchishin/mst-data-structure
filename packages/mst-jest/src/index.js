import * as mobxStateTree from 'mobx-state-tree';
import {onPatch, onSnapshot, getSnapshot } from 'mobx-state-tree';

export {
    protect,
    unprotect,
    isProtected,
} from 'mobx-state-tree';

export const createMSTSnapshot = (model, ...options) => {
    expect(getSnapshot(model)).toMatchSnapshot(...options);
};

export const getEnvMock = (value = {}) => {
    jest.spyOn(mobxStateTree, 'getEnv').mockReturnValue(value);
};

export const getRootMock = (value = {}) => {
    jest.spyOn(mobxStateTree, 'getRoot').mockReturnValue(value);
};

export const mockMiddleware = (call, next) => next(call);

export const onPatchesAggregation = (model, ...options) => {
    const patches = [];

    const disposer = onPatch(model, patch => {
        patches.push(patch);
    });

    return () => {
        expect(patches).toMatchSnapshot(...options);
        disposer();
    };
};

export const onSnapshotAggregation = (model, ...options) => {
    const snapshots = [];

    const disposer = onSnapshot(model, patch => {
        snapshots.push(patch);
    });

    return () => {
        expect(snapshots).toMatchSnapshot(...options);
        disposer();
    };
};
