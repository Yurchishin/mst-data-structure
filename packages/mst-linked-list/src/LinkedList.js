import { types } from 'mobx-state-tree';
import { values } from 'mobx';

export const optionalMap = (Type, optionalValues) => types.optional(types.map(Type), {}, optionalValues)
export const optionalNull = (Type, optionalValues) => types.optional(types.maybeNull(Type), null, optionalValues)

const LinkedList = (name = '', Node) => {
    const LinkedListModel =  types
        .model(`${name}LinkedList`, {
            head: optionalNull(types.reference(Node)),
            tail: optionalNull(types.reference(Node)),
            nodes: optionalMap(Node),
        })
        .views(self => ({
            getNode(id) {
                return self.nodes.get(id) || null;
            },
            get(id) {
                return self.nodes.get(id).value || null;
            },
            has(id) {
                return self.nodes.has(id);
            },
            get values() {
                return values(self.nodes)
            },
            get entries() {
                return self.values.map(node => [node.id, node.value]);
            },
            get array() {
                return self.values.map(node => node.value);
            },
            get isEmpty() {
              return self.nodes.size === 0
            },
            toString(callback) {
                return self.array.map(node => node.toString(callback)).toString();
            },
        }))
        .actions(self => {
            const prepend = (id, value) => {
                if(!self.head) {
                    self.nodes.set(id, {
                        id,
                        value,
                        next: null,
                    });
                }
                else {
                    self.nodes.set(id, {
                        id,
                        value,
                        next: self.head.id,
                    });
                }

                self.head = id;

                if (!self.tail) {
                    self.tail = id;
                }

                return self;
            };

            const append = (id, value) => {
                self.nodes.set(id, {
                    id,
                    value,
                    next: null,
                });

                if (!self.head) {
                    self.head = id;
                    self.tail = id;

                    return self;
                }

                self.tail.setNext(id);
                self.tail = id;

                return self;
            };

            const remove = id => {
                const deletedNode = self.nodes.get(id);

                if (!deletedNode) return self;
                if (!self.head) return self;

                if (self.head === deletedNode) {
                    if (self.head === self.tail) {
                        self.head = null;
                        self.tail = null;
                    } else {
                        self.head = self.head.next;
                    }
                } else {
                    let currentNode = self.head;

                    if (currentNode) {
                        while (currentNode.next) {
                            if (currentNode.next === deletedNode) currentNode.setNext(currentNode.next.next);
                            else currentNode = currentNode.next;
                        }
                    }

                    if (self.tail === deletedNode) self.tail = currentNode;
                }

                self.nodes.delete(id);

                return self;
            };

            const fromArray = values => {
                values.forEach(([id, value]) => self.append(id, value));

                return self;
            };

            const deleteTail = () => {
                if(!self.head) return self;
                const id = self.tail.id

                if (self.head === self.tail) {
                    self.head = null;
                    self.tail = null;

                    self.nodes.delete(id);

                    return self;
                }

                // Rewind to the last node and delete "next" link for the node before the last one.
                let currentNode = self.head;
                while (currentNode.next) {
                    if (!currentNode.next.next) currentNode.setNext(null);
                    else currentNode = currentNode.next;
                }

                self.tail = currentNode;

                self.nodes.delete(id);

                return self;
            }

            const deleteHead = () => {
                if (!self.head) return self

                const id = self.head.id

                if (self.head.next) self.head = self.head.next;
                else {
                    self.head = null;
                    self.tail = null;
                }

                self.nodes.delete(id);

                return self;
            }

            const reverse = () => {
                let currNode = self.head;
                let prevNode = null;
                let nextNode = null;

                while (currNode) {
                    nextNode = currNode.next;
                    currNode.setNext(prevNode);

                    prevNode = currNode;
                    currNode = nextNode;
                }

                self.tail = self.head;
                self.head = prevNode;

                return self;
            }

            return {
                prepend,
                append,
                remove,
                fromArray,
                deleteHead,
                deleteTail,
                reverse,
            };
        });

    return types.optional(LinkedListModel, {})
};

export default LinkedList;