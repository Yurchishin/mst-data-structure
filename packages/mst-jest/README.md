## `mst-jest`

#### API

##### `createMSTSnapshot(node)` - created jest snapshot from model
##### `getEnvMock({...})` - mocked for `getEnv`'
##### `getRootMock({...})` - mocked for `getRoot`'
##### `onPatchesAggregation(node)` - aggregate all model patches, and take a jest snapshot'
##### `onSnapshotAggregation(node)` - aggregate all model snapshots, and take a jest snapshot'

#### Usage
```js
import { types } from "mobx-state-tree"
import { createMSTSnapshot, onPatchesAggregation, onSnapshotAggregation } from "@mst-ds/mst-jest"

describe('Tests', () => {
    it('snapshot test', () => {
        const rootStore = types.model({
            ...,
        })
        .actions(() => {...});

        createMSTSnapshot(list) // create jest snapshot
    });
it('snapshots test', () => {
        const rootStore = types.model({
            ...,
        })
        .actions(() => {...});

        const disposer = onPatchesAggregation(list)

        rootStore.action1();
        rootStore.action2();
        rootStore.action3();

        disposer() // create jest snapshot
    });
    it('patches test', () => {
        const rootStore = types.model({
            ...,
        })
        .actions(() => {...});

        const disposer = onSnapshotAggregation(list)

        rootStore.action1();
        rootStore.action2();
        rootStore.action3();

        disposer() // create jest snapshot
    });
});
```