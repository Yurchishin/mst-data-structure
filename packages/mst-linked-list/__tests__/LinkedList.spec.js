import { types } from 'mobx-state-tree';
import { onSnapshotAggregation, createMSTSnapshot } from '@mst/mst-jest';
import LinkedList from '../src/index';

describe('LinkedList', () => {
    describe('model', () => {
        it('create(int)', () => {
            const list = LinkedList('Int', types.integer).create();

            createMSTSnapshot(list)

            expect(list.head).toBeNull();
            expect(list.tail).toBeNull();
        });
        it('create in model(int)', () => {
            const rootStore = types.model('RootStore', {
                numbers: LinkedList('Int', types.integer),
            }).create()

            createMSTSnapshot(rootStore)
        });
        it('create in model(Todo)', () => {
            const Todo = types.model('Todo', {
                name: types.string,
                active: types.boolean,
            })

            const rootStore = types.model('RootStore', {
                todos: LinkedList('Todo', Todo),
            }).create()

            rootStore.todos.append('1', {
                name: 'name1',
                active: false,
            })
            .append('2', {
                name: 'name2',
                active: true,
            })
            .append('3', {
                name: 'name2',
                active: false,
            })

            createMSTSnapshot(rootStore)
        });
    })
    describe('actions', () => {
        it('prepend', () => {
            const list = LinkedList('Int', types.integer).create();

            const disposer = onSnapshotAggregation(list)

            expect(list.head).toBeNull();
            expect(list.tail).toBeNull();

            list.prepend('3', 3);

            expect(list.head.id).toBe('3');
            expect(list.head.value).toBe(3);
            expect(list.tail.id).toBe('3');
            expect(list.tail.value).toBe(3);

            list.prepend('2', 2).prepend('1', 1);

            const { head } = list;
            expect(head.id).toBe('1');
            expect(head.value).toBe(1);
            expect(list.nodes.size).toBe(3);

            const node2 = head.next;
            expect(node2.id).toBe('2');
            expect(node2.value).toBe(2);

            const node3 = node2.next;
            expect(node3.id).toBe('3');
            expect(node3.value).toBe(3);

            const { tail } = list;
            expect(tail.id).toBe('3');
            expect(tail.value).toBe(3);

            disposer()
        });

        it('append', () => {
            const list = LinkedList('Int', types.integer).create();

            const disposer = onSnapshotAggregation(list)

            expect(list.head).toBeNull();
            expect(list.tail).toBeNull();

            list.append('1', 1);

            expect(list.head.id).toBe('1');
            expect(list.head.value).toBe(1);
            expect(list.tail.id).toBe('1');
            expect(list.tail.value).toBe(1);

            list.append('2', 2).append('3', 3);

            const { head } = list;
            expect(head.id).toBe('1');
            expect(head.value).toBe(1);

            const node2 = head.next;
            expect(node2.id).toBe('2');
            expect(node2.value).toBe(2);

            const node3 = node2.next;
            expect(node3.id).toBe('3');
            expect(node3.value).toBe(3);

            const { tail } = list;
            expect(tail.id).toBe('3');
            expect(tail.value).toBe(3);

            disposer()
        });

        it('remove head and tail', () => {
            const list = LinkedList('Int', types.integer).create();

            const disposer = onSnapshotAggregation(list)

            expect(list.head).toBeNull();
            expect(list.tail).toBeNull();
            expect(list.nodes.size).toBe(0);

            list.append('1', 1);

            expect(list.head.id).toBe('1');
            expect(list.head.value).toBe(1);
            expect(list.tail.id).toBe('1');
            expect(list.tail.value).toBe(1);
            expect(list.nodes.size).toBe(1);

            list.remove('1');

            expect(list.head).toBeNull();
            expect(list.tail).toBeNull();
            expect(list.nodes.size).toBe(0);

            disposer()
        });

        it('remove head', () => {
            const list = LinkedList('Int', types.integer).create();

            const disposer = onSnapshotAggregation(list)

            expect(list.head).toBeNull();
            expect(list.tail).toBeNull();
            expect(list.nodes.size).toBe(0);

            list.append('1', 1).append('2', 2);

            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('2');
            expect(list.nodes.size).toBe(2);

            list.remove('1');

            expect(list.head.id).toBe('2');
            expect(list.tail.id).toBe('2');
            expect(list.nodes.size).toBe(1);

            list.prepend('1', 1).append('3', 3);

            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('3');
            expect(list.nodes.size).toBe(3);

            list.remove('1');

            expect(list.head.id).toBe('2');
            expect(list.tail.id).toBe('3');
            expect(list.nodes.size).toBe(2);

            disposer()
        });

        it('remove middle', () => {
            const list = LinkedList('Int', types.integer).create();

            const disposer = onSnapshotAggregation(list)

            expect(list.head).toBeNull();
            expect(list.tail).toBeNull();
            expect(list.nodes.size).toBe(0);

            list
                .append('1', 1)
                .append('2', 2)
                .append('3', 3);

            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('3');
            expect(list.nodes.size).toBe(3);

            list.remove('2');

            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('3');
            expect(list.nodes.size).toBe(2);

            disposer()
        });

        it('remove tail', () => {
            const list = LinkedList('Int', types.integer).create();

            const disposer = onSnapshotAggregation(list)

            expect(list.head).toBeNull();
            expect(list.tail).toBeNull();
            expect(list.nodes.size).toBe(0);

            list
                .append('1', 1)
                .append('2', 2)
                .append('3', 3);

            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('3');
            expect(list.nodes.size).toBe(3);

            list.remove('3');
            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('2');
            expect(list.nodes.size).toBe(2);

            list.remove('2');
            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('1');
            expect(list.nodes.size).toBe(1);

            list.remove('1');
            expect(list.head).toBeNull();
            expect(list.tail).toBeNull();
            expect(list.nodes.size).toBe(0);

            disposer()
        });

        it('fromArray', () => {
            const list = LinkedList('Int', types.integer).create();

            const disposer = onSnapshotAggregation(list)

            expect(list.head).toBeNull();
            expect(list.tail).toBeNull();
            expect(list.nodes.size).toBe(0);

            list.fromArray([['1', 1]]);
            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('1');
            expect(list.nodes.size).toBe(1);

            list.fromArray([['2', 2], ['3', 3]]);
            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('3');
            expect(list.nodes.size).toBe(3);

            disposer()
        });

        it('deleteTail', () => {
            const list = LinkedList('Int', types.integer).create();

            const disposer = onSnapshotAggregation(list)

            list.append('1', 1);
            list.deleteTail();
            expect(list.head).toBeNull();
            expect(list.tail).toBeNull();
            expect(list.nodes.size).toBe(0);

            list.append('1', 1).append('2', 2);
            list.deleteTail();
            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('1');
            expect(list.nodes.size).toBe(1);

            list.append('2', 2).append('3', 3);
            list.deleteTail();
            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('2');
            expect(list.nodes.size).toBe(2);

            disposer()
        });

        it('deleteHead', () => {
            const list = LinkedList('Int', types.integer).create();

            const disposer = onSnapshotAggregation(list)

            list.append('1', 1);
            list.deleteHead();
            expect(list.head).toBeNull();
            expect(list.tail).toBeNull();
            expect(list.nodes.size).toBe(0);

            list.append('2', 2).append('1', 1);
            list.deleteHead();
            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('1');
            expect(list.nodes.size).toBe(1);

            list.append('2', 2).append('3', 3);
            list.deleteHead();
            expect(list.head.id).toBe('2');
            expect(list.tail.id).toBe('3');
            expect(list.nodes.size).toBe(2);

            disposer()
        });

        it('reverse', () => {
            const list = LinkedList('Int', types.integer).create();

            const disposer = onSnapshotAggregation(list)

            list.append('1', 1);
            list.reverse();
            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('1');

            list.append('2', 2)
            list.reverse();
            expect(list.head.id).toBe('2');
            expect(list.tail.id).toBe('1');

            list.reverse();
            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('2');

            list.append('3', 3)
            list.reverse();
            expect(list.head.id).toBe('3');
            expect(list.tail.id).toBe('1');

            list.reverse();
            expect(list.head.id).toBe('1');
            expect(list.tail.id).toBe('3');

            disposer()
        });
    });
});
