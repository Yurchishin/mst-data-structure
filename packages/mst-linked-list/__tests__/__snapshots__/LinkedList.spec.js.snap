// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkedList actions append 1`] = `
Array [
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
]
`;

exports[`LinkedList actions deleteHead 1`] = `
Array [
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": null,
    "nodes": Object {},
    "tail": null,
  },
  Object {
    "head": "2",
    "nodes": Object {
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "2",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "1",
        "value": 2,
      },
    },
    "tail": "1",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
  Object {
    "head": "2",
    "nodes": Object {
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
]
`;

exports[`LinkedList actions deleteTail 1`] = `
Array [
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": null,
    "nodes": Object {},
    "tail": null,
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
]
`;

exports[`LinkedList actions fromArray 1`] = `
Array [
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
]
`;

exports[`LinkedList actions prepend 1`] = `
Array [
  Object {
    "head": "3",
    "nodes": Object {
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
  Object {
    "head": "2",
    "nodes": Object {
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
]
`;

exports[`LinkedList actions remove head 1`] = `
Array [
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "2",
    "nodes": Object {
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
  Object {
    "head": "2",
    "nodes": Object {
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
]
`;

exports[`LinkedList actions remove head and tail 1`] = `
Array [
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": null,
    "nodes": Object {},
    "tail": null,
  },
]
`;

exports[`LinkedList actions remove middle 1`] = `
Array [
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "3",
        "value": 1,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
]
`;

exports[`LinkedList actions remove tail 1`] = `
Array [
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": null,
    "nodes": Object {},
    "tail": null,
  },
]
`;

exports[`LinkedList actions reverse 1`] = `
Array [
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
    },
    "tail": "1",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "2",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "1",
        "value": 2,
      },
    },
    "tail": "1",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": null,
        "value": 2,
      },
    },
    "tail": "2",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
  Object {
    "head": "3",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": null,
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "1",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": "2",
        "value": 3,
      },
    },
    "tail": "1",
  },
  Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": 1,
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": 2,
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": 3,
      },
    },
    "tail": "3",
  },
]
`;

exports[`LinkedList model create in model(Todo) 1`] = `
Object {
  "todos": Object {
    "head": "1",
    "nodes": Object {
      "1": Object {
        "id": "1",
        "next": "2",
        "value": Object {
          "active": false,
          "name": "name1",
        },
      },
      "2": Object {
        "id": "2",
        "next": "3",
        "value": Object {
          "active": true,
          "name": "name2",
        },
      },
      "3": Object {
        "id": "3",
        "next": null,
        "value": Object {
          "active": false,
          "name": "name2",
        },
      },
    },
    "tail": "3",
  },
}
`;

exports[`LinkedList model create in model(int) 1`] = `
Object {
  "numbers": Object {
    "head": null,
    "nodes": Object {},
    "tail": null,
  },
}
`;

exports[`LinkedList model create(int) 1`] = `
Object {
  "head": null,
  "nodes": Object {},
  "tail": null,
}
`;
